### ISSUE TITLE
Default bundler version for ruby 2.6.x installations

### ISSUE THREAD

**Issue by pclalv** [rubygems/rubygems#2842]
I have applications that require bundler 2.0.2 and ruby 2.6.3. Ruby 2.6.3 comes with bundler 1.17.2, so I need to install bundler 2.0.2; after I install bundler 2.0.2, I have two version of bundler. Invoking `gem uninstall bundler --version 1.17.2` doesn't seem to be able to uninstall the bundler provided by ruby 2.6.x, so at this point I need to be able to guarantee that I use the correct one. **Is there some means of selecting the bundler that I want (2.0.2) so that I can avoid running the bundler that I don't want (1.17.2)?** I know that I can request that bundler 2.0.2 be used by invoking, for example, `bundle _2.0.2_ exec ...`, but that's not what I want - I don't want to have to do that every time I need to invoke bundler, and it's impractical in my containerized ruby environments . I would like to avoid any kind of homegrown shim (whether shell alias, shell function, or shell script) that tries to figure out what version of bundler should be used. I see too that there's a `--default` option that can be passed to gem install, but it's not clear to me how that might help me. This is what I saw after running `gem install bundler --version 2.0.2 --default` with a fresh ruby 2.6.3 installation:

```
$ gem list bundler

*** LOCAL GEMS ***

bundler (default: 2.0.2, default: 1.17.2)
```

For now, it seems like the best workaround is to programmatically edit the bundler executable provided by RubyGems so that my desired bundler version is always assigned to version.  

I was hoping that RubyGems would provide some means of selecting a default bundler version, say by an environment variable `BUNDLER_DEFAULT_VERSION`. Does anything like that exist? If not, would it be reasonable for RubyGems to provide such a mechanism? Is there a bug in any of the behavior I've observed? Or is all of this behavior intended?

**Comment by simi** [comment]
Default is probably something different then expect.

```
        --default                    Add the gem's full specification to                       
                                     specifications/default and extract only its bin           

```

Anyway I was able to broke my rubygems installation by:

```
gem install bundler -v 2.0.2 --default
gem uninstall bundler # uninstall default one
gem uninstall bundler # uninstall 2.0.2
gem install bundler

gem list bundler

*** LOCAL GEMS ***

bundler (default: 2.0.2, default: 1.17.2)

bundle -v
Traceback (most recent call last):
        1: from /home/retro/.gem/ruby/2.6.3/bin/bundle:23:in `<main>'
/home/retro/.gem/ruby/2.6.3/bin/bundle:23:in `load': cannot load such file -- /home/retro/.rubies/ruby-2.6.3/lib/ruby/gems/2.6.0/gems/bundler-2.0.2/exe/bundle (LoadError)
gem install bundler -v 2.0.2
bundle -v
Traceback (most recent call last):
        1: from /home/retro/.gem/ruby/2.6.3/bin/bundle:23:in `<main>'
/home/retro/.gem/ruby/2.6.3/bin/bundle:23:in `load': cannot load such file -- /home/retro/.rubies/ruby-2.6.3/lib/ruby/gems/2.6.0/gems/bundler-2.0.2/exe/bundle (LoadError)
```

**Comment by hsbt** [comment]
```--default``` option is broken. It's the historical reason and no one cares about it.

@pclalv You never uninstall default gems from ruby installation and don't need to uninstall 1.17.x for uses bundler-2.0.x.

**Comment by deivid-rodriguez** [comment]
@hsbt Does that mean that you're no longer interested in https://github.com/rubygems/rubygems/pull/2166?

I think fixing/improving the `--default` flag would be interesting not only so that it doesn't break any installations of users trying it, but because it could be reused by ruby-core's installer, which as far as I recall uses quite a lot of custom code just to get the default gems properly installed.

@simi The error you saw should be fixed by https://github.com/rubygems/rubygems/pull/2649, which is scheduled for 3.0.5.

**Comment by deivid-rodriguez** [comment]
Also, the fact that you got no feedback whasoever when running `gem uninstall bundler` was fixed by https://github.com/rubygems/rubygems/pull/2707. Once we release that PR, you should see an error message telling you that bundler is a default gem and can't be uninstalled.

@hsbt Can we also schedule that PR for 3.0.5?

**Comment by pclalv** [comment]
@hsbt @deivid-rodriguez i appreciate your comments, but i don't understand how they speak to the issue that i'm trying to point out. i apologize if my first comment was unclear.

to put things in simpler terms, imagine you've got a standard ruby 2.6.3 project directory with a `Gemfile` and `Gemfile.lock` and whatever else. at the bottom of `Gemfile.lock` it says:
```
BUNDLED WITH
   2.0.2
```

that's the version of bundler i want to use. ideally, there would be some mechanism for automatically choosing the correct bundler version. right now, we've hacked the rubygems binstup for bundler by overriding the `version ="0.a"` assignment with whatever bundler version we actually want, which is a little hacky. i was hoping that rubygems might be able to provide a better mechanism, even if it's to pull from an env var called `RUBYGEMS_BUNLDER_VERSION`. the env var approach would definitely work for the containers that i'm working with.

**Comment by deivid-rodriguez** [comment]
@pclalv Sorry, I only replied to the last comment by @hsbt but didn't actually read your issue. From a first look, it seems to me that rubygems should be actually choosing the right bundler version the way you expect, so if it's not doing that, it seems like a bug to me... :/

I'll have a closer look as soon as I can :+1:.

**Comment by pclalv** [comment]
@deivid-rodriguez no worries and no pressure; i know responding to these issues takes time.

**Comment by deivid-rodriguez** [comment]
@pclalv I had actually forgotten about this... I read the issue now, and yeah, I think rubygems should take care of doing what you want automatically.

If you are inside an application, rubygems should read your `Gemfile.lock` file, and use the bundler version specified in the `BUNDLED WITH` section.

If you are not inside an application, rubygems will run the highest bundler version installed on your system (unless you use the `bundle _<version>_` trick).

When I say "rubygems", I refer to the rubygems binstub that gets run when you run your the `bundle` executable installed on your system by rubygems.

If you're not seeing this behavior, it would be great if you managed to provide some reproduction steps, so I can have a closer look.

Thanks!

**Comment by Nakilon** [comment]
Thanks, now my bundler is broken too.
I had no such problems with 2.4 and 2.5.

**Comment by deivid-rodriguez** [comment]
Closing this since no feedback was provided. Thanks!
